src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -603,18 +599,47 @@ ModifyQuerySupported(Query *queryTree, Query *originalQuery, bool multiShardQuer
 		}
 	}
 
-	/* reject queries which include CommonTableExpr */
+	/* reject queries which include CommonTableExpr which aren't routable */
 	if (queryTree->cteList != NIL)
 	{
-		return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
-							 "common table expressions are not supported in distributed "
-							 "modifications",
-							 NULL, NULL);
+		ListCell *cteCell = NULL;
+		DeferredErrorMessage *cteError = NULL;
+
+		foreach(cteCell, queryTree->cteList)
+		{
+			CommonTableExpr *cte = (CommonTableExpr *) lfirst(cteCell);
+			Query *cteQuery = (Query *) cte->ctequery;
+			bool requiresMasterEvaluation;
+
+			if (cteQuery->commandType != CMD_SELECT)
+			{
+				return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
+									 "Router planner doesn't support non-select common table expressions.",
+									 NULL, NULL);
+			}
+
+			requiresMasterEvaluation = RequiresMasterEvaluation(cteQuery);
+			if (requiresMasterEvaluation)
+			{
+				return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
+									 "Router planner doesn't support common table expressions requiring master evaluation.",
+									 NULL, NULL);
+			}
+
+			/* TODO can we reduce redundancy here by directly calling PlanRouterQuery? */
+			cteError = CreateRouterPlan(cteQuery, cteQuery,
```

Comment:
It's a little odd that we're doing all of `CreateRouterPlan` here rather than just the error checks. Maybe we could just do `MultiRouterPlannableQuery` after we move `ErrorIfQueryHasModifyingCTE`?

Reply:
It was running into "Complex subqueries and CTEs" type of errors without checking if PlanRouterQuery wasn't going to bail either. Might be related to range table scan seeming to be skipping CTEs

My other understanding is that I need information from PlanRouterQuery in order to make sure the CTE belongs to the same shard (or, more broadly, worker) as the query. Need to test `with cte1 as (query-using-shard-1), cte2 as (query-using-shard-2) select * from (query-using-ctes)`

Reply:
`with cte1 as (query-using-shard-1), cte2 as (query-using-shard-2) select * from (query-using-ctes)` is only router plannable if shard 1 and 2 are co-located shards from different tables. Otherwise we should be using recursive planning like we do now.

My expectation is that `PlanRouterQuery` only needs to be called once for the whole query, since it checks whether all of the shards that are used in the query belong to the same worker.